integrations/weights_bias/src/haystack_integrations/tracing/weave/tracer.py

```diff
@@ -0,0 +1,158 @@
+
+        :param key: The tag key.
+        :param value: The tag value.
+        """
+        coerced_value = value
+        self._attributes[key] = coerced_value
+
+    def set_tags(self, tags: Dict[str, Any]) -> None:
+        for k, v in tags.items():
+            self.set_tag(k, v)
+
+    def raw_span(self) -> Any:
+        """Access to the underlying Weave Call object."""
+        return self._call
+
+    def get_correlation_data_for_logs(self) -> Dict[str, Any]:
+        """Correlation data for logging."""
+        if not self._call:
+            return {}
+        return {
+            "weave.call_id": self._call.id,
+            "weave.run_id": getattr(self._call, "run_id", ""),
+        }
+
+    def set_call(self, call: "weave.Call") -> None:
+        self._call = call
+
+    def get_attributes(self) -> Dict[str, Any]:
+        return self._attributes
+
+
+def create_call(
+    attributes: Dict,
+    client: WeaveClient,
+    parent_span: Union[WeaveSpan, None],
+    operation_name: str,
+) -> Call:
+    comp_name = attributes.pop("haystack.component.name", "")
+    comp_type = attributes.pop("haystack.component.type", "")
+    comp_input = attributes.pop("haystack.component.input", {})
+    call = client.create_call(
+        op=operation_name,
+        inputs={key: coerce_tag_value(value) for key, value in comp_input.items()},
+        attributes={key: coerce_tag_value(value) for key, value in attributes.items()},
+        display_name=f"{comp_name}[{comp_type}].run",
+        parent=parent_span.raw_span() if parent_span else None,
+    )
+
+    return call
+
+
+class WeaveTracer(Tracer):
+    """A simple bridge between Haystack's Tracer interface and Weave."""
+
+    def __init__(self, project_name: str) -> None:
+        self._client = weave.init(project_name)
+        self._current_span: Optional[WeaveSpan] = None
+
+    @contextlib.contextmanager
+    def trace(
+        self,
+        operation_name: str,
+        tags: Optional[Dict[str, Any]] = None,
+        parent_span: Optional[WeaveSpan] = None,
+    ) -> Iterator[WeaveSpan]:
+        """Create a new trace span."""
+        # We need to defer call creation for components as a Call in Weave can't be updated
+        # but the content tags are only set on the Span at a later stage. To get the inputs on
+        # call creation, we need to create the call after we yield the span.
+        if operation_name == "haystack.component.run":
+            span = WeaveSpan(
+                parent=parent_span.raw_span() if parent_span else None,
+                operation=operation_name,
+            )
+        else:
+            call = self._client.create_call(
+                op=operation_name,
+                inputs=tags,
+                parent=parent_span.raw_span() if parent_span else None,
+            )
+            span = WeaveSpan(call=call)
+        self._current_span = span
+
+        if tags:
+            span.set_tags(tags)
+
```

Comment:
The following part seems to me a bit hard to read and understand, mostly because of nested logic.
It would be great if it could be simplified a bit.

Reply:
I agree. I'm trying to understand this code a bit more in detail and see how to improve it - although I'm opening this PR the code was made by someone else

Reply:
I did some refactoring for the trace method, following also on an extended docstring, explaining better what's happening in the trace method and why the two approaches depending on the type of operation we want to trace